In [1]:
import random

from IPython.display import IFrame
IFrame("Blatt05.pdf", width=1000, height=500)

In [2]:
def gen_prime_factors(n):
    i = 2
    factors = []
    while i * i <= n:
        if n % i:
            i += 1
        else:
            n //= i
            factors.append(i)
    if n > 1:
        factors.append(n)
    return list(set([(x, factors.count(x)) for x in factors]))


def ggT(a,b):
    if b == 0: return a
    return ggT(b, a % b)


def fermat_test_carmichael_prob(n):
    divider = list(range(2,n))
    prime_factors = [x[0] for x in gen_prime_factors(n)]
    for i in range(n):
        for x in prime_factors:
            if ggT(i,x) > 1:
                try:
                    divider.remove(i)
                except ValueError:
                    pass
                    #print("Value " + str(i) + " already removed.")
    return len(divider) / n


print(fermat_test_carmichael_prob(561))
print(fermat_test_carmichael_prob(1105))

0.5686274509803921
0.6941176470588235


561 = 3\*11\*17\
1105 = 5\*13\*19\
forall a el zn^*: a^n-1 = 1 mod n\
ggt(a,n)=1\
phi(n)-1/#{2,...,n-1} = phi(n)-1/n-2\
phi(561)=2\*10\*16=320 => 320-1/561-2=0.57\
phi(1105)=4\*12\*18=864 => 864-1/1105-2 = 0.78

# 5.2

Euler-Kriterium: a el Zn^\* quadr. Rest gdw. a^(n-1)/2 = 1 mod n\
Primitivwurzelkriterium: a el Zn^\* PW gdw. a^(phi(n)/q) != 1 mod n forall q Primzahl: q | phi(n)

a) $a^{(n-1)/2}\ ≡\ -1\ mod\ n$\
b) $a^{(n-1)/2}\ ≡\ +1\ mod\ n$

Es folgt aus dem Euler-Kriterium, dass b) ein quadratischer Rest ist und daher keine Primitivwurzel sein kann (da 2 teiler von phi(n) ist PWK nicht erfüllt) und a) ein quadratischer Nicht-Rest, da ungleich 1 (EK) und ist PW (solange für andere Primteiler q von a^phi(n)/q auch != 1 ist)


# 5.3

Diffie-Hellman -> ElGamal:\
Gegeben: $n$, $a$, $x'=a^x\ mod\ n$, $y'=a^{y}\ mod\ n$, $m'=a^{xy} m\ mod\ n$\
Ziel: Bestimme m\
1: Frage Orakel nach $z=a^{xy}\ mod\ n$ und berechne inverses mod n\
2: Berechne $m' z^{-1} = m a^{xy} (a^{xy})^{-1} = m\ mod\ n$

ElGamal -> Diffie-Hellman:\
Gegeben: $n$, $a$, $x'=a^x\ mod\ n$, $y'=a^{y}\ mod\ n$\
Ziel: Bestimme $z=a^{xy}\ mod\ n$\
1: Wähle verschlüsselte Nachricht $m'$ mit $m'=a^{xy} m\ mod\ n$\
2: ElGamal gibt uns $m$\
3: Berechne inverses zu $m$ und berechne $m' * m^{-1} = a^{xy} * m * m^{-1} = a^{xy}\ mod\ n$

# 5.4

#### a)
$h∈P$, da $h(x)=g(f(x))$ die Komposition von zwei polynomialzeit-berechenbaren Funktionen $f,g$ ist\
$h^{-1}∉P$, da sonst $x$ so gefunden werden kann, dass $g(f(x))=y$. Da $f$ eine Einwegfunktion ist, ist $f^{-1}∉P$, mit $g$ ebenso, also ist $h^{-1}∉P$

#### b)
\
$f$ ist Einwegfunktion mit Falltür gdw. $f$ ist Einwegfkt. & Ex. eff. Alg. M & forall n Ex. Schlüssel k_n: M(y,k_n)=$f^{-1}(y)$\
Sei M_f eff. Alg. für $f$ & k_n Schlüssel: M_f(y,k_n)=$f^{-1}(y)$\
Sei M_g eff. Alg. für $g$ & k_n' Schlüssel: M_g(y,k_n')=$g^{-1}(y)$\
$h^{-1}(y)=(f^{-1}°g^{-1})(y)=f^{-1}(g^{-1}(y))=F^{-1}(M_g(y,k_n'))=M_f(M_g(y,k_n'),k_n)$

# 5.5

Für $n$ gibt es $x,y: x^2≡y^2\ mod\ n$ und $x≢∓y\ mod\ n$\
ZZ: $n$ ist keine Primzahl und ggT($x-y$,$n$) ist nicht-trivialer Faktor von n\
$n$ ist Teiler von $x^2-y^2\ =\ (x-y)(y+x)$\
Da $x≢∓y\ mod\ n$ folgt $x-y≢0\ mod\ n$ und $x+y≢0\ mod\ n$, d.h. weder $x-y$ noch $x+y$ ist durch n teilbar\
Da $n\ |\ x^2-y^2\ =\ (x-y)(y+x)$ und $x-y$ und $x+y$ $n$ nicht teilen ist n keine Primzahl und muss zusammengesetzt sein, da es mindestens einen weiteren Teiler außer 1 und $n$ gibt\
daraus folgt auch, dass der ggT($x-y$,$n$) ist nicht-trivialer Faktor von n ist

# Übung:

Ang. d=ggT(x-y,n)=n => x-y = 0 mod n => x = y mod n Error\
Ang. d=1, [a|b\*c & ggT(a,b) => a|c]\
x^2-y^2=0 mod n => n|x^2-y^2\
x^2-y^2 = (x+y)(x-y) => n|x+y => x+y=0 mod n => x=-y mod n Error\
=> d ist nicht trivialer Faktor von n => n keine Primzahl